@spinningthelock thanks for this

## What I verified

**Integrity.** Downloaded `gpsd-3.27.5.tar.xz` from savannah, sha256 matches `sha256-3Epiutg1KCuueIdyvHzI+L7Ex6SOjc7rN0d6iQkcRlY=` as declared. All three overlay hashes in `source.json` recompute to exactly what is checked in.

**The `_GNU_SOURCE` fix is real.** I compiled the full `libgps` srcs list under gcc 13 / glibc with the old header versus the new one:

- old: `libgps/gpsutils.c:892:14: warning: implicit declaration of function 'strptime'`
- new: clean

At `gpsutils.c:892` the code is `char *dp = strptime(isotime, ...)` followed by `'.' == *dp`. With no declaration, glibc's `strptime` is assumed to return `int`, so the pointer is truncated to 32 bits then dereferenced. That is the segfault and it sits in `iso8601_to_timespec`, which every gpsd JSON time field goes through. Nice catch. This matches upstream `SConscript:1029-1031`.

**It links.** I built the branch as a local registry with Bazel 9.2.0 on macOS arm64 and added a C++ test that links `@libgps//:libgpsmm` and exercises the constructor, `is_open`, `data`, `clear_fix`, `enable_debug` and the destructor. That plus `@libgps//:libgps_test` both pass. Every symbol `libgpsmm.cpp` references is present in the srcs list, inside `SOCKET_EXPORT_ENABLE` which is defined.

## Changes requested

**1. Rename `3.27.5.bcr.0` to `3.27.5.bcr.1`.** `docs/README.md` gives `1.2.3.bcr.1` as the convention for the first BCR-only fix and 879 module versions in the registry use `.bcr.1` against 14 using `.bcr.0`. The registry is add-only so this cannot be corrected after merge. Rename the directory, update `version` in `MODULE.bazel` and the entry in `metadata.json`.

**2. Add a C++ test target.** `presubmit.yml` only *builds* `@libgps//:libgpsmm`. Building a `cc_library` compiles and archives but does not link, so an undefined symbol in the C++ wrapper would pass CI unnoticed. I confirmed linking by hand, but the module should carry that guarantee itself, particularly for the next gpsd bump. Please add an `overlay/tests/libgpsmm_test.cc` with a `cc_test` depending on `:libgpsmm`, list it in both `build_targets` and `test_targets`, then re-run `bazel run //tools:update_integrity`.

## Nits

- `cc_library(name = "gpsmm")` sets `visibility = ["//visibility:public"]`, but `:gps` does not. The pattern in this file is a private target plus a public alias, so please drop `visibility` from the `gpsmm` rule and let `:libgpsmm` be the single public entry point.
- The `gpsmm` rule puts `hdrs` before `srcs`. Every other rule in the file uses `srcs` then `hdrs`.
- `_GNU_SOURCE` is unconditional here while upstream scopes it to the Linux branch. Harmless in practice (I confirmed the macOS build is fine) plus the file already applies `_DARWIN_C_SOURCE` unconditionally, so it is internally consistent. Optional: guard it with `#if defined(__linux__)` to mirror upstream exactly.
- The `rules_cc` bump from 0.2.18 to 0.2.22 is unrelated to the stated purpose of the PR. It works fine, just worth a line in the description.
- `modules/libgps/README.md` documents the overlay but not the exported targets. Could be worth mentioning `@libgps//:libgps` (C) and `@libgps//:libgpsmm` (C++).

--- a/modules/libgps/3.27.5.bcr.1/MODULE.bazel
+++ b/modules/libgps/3.27.5.bcr.1/MODULE.bazel
@@ -1,0 +1,11 @@
+"""https://gpsd.io/"""
+
+module(
+    name = "libgps",
+    version = "3.27.5.bcr.1",
+    compatibility_level = 1,
+    bazel_compatibility = [">=7.2.1"],
+)
+
+bazel_dep(name = "rules_cc", version = "0.2.22")
+bazel_dep(name = "rules_python", version = "1.9.0")

--- a/modules/libgps/3.27.5.bcr.1/overlay/BUILD.bazel
+++ b/modules/libgps/3.27.5.bcr.1/overlay/BUILD.bazel
@@ -1,0 +1,123 @@
+load("@rules_cc//cc:defs.bzl", "cc_library", "cc_test")
+load("@rules_python//python:py_binary.bzl", "py_binary")
+
+py_binary(
+    name = "maskaudit",
+    srcs = ["maskaudit.py"],
+    main = "maskaudit.py",
+)
+
+py_binary(
+    name = "jsongen",
+    srcs = ["libgps/jsongen.py"],
+    main = "libgps/jsongen.py",
+)
+
+genrule(
+    name = "gps_maskdump_src",
+    srcs = [
+        "gpsctl.c",
+        "include/gps.h",
+        "include/gpsd.h",
+        "libgps/libgps_core.c",
+        "libgps/libgps_json.c",
+    ],
+    tools = [":maskaudit"],
+    outs = ["libgps/gps_maskdump.c"],
+    cmd = "mkdir -p $(@D) && ROOT=$$(pwd) && OUT=$$(pwd)/$@ && cd $$(dirname $(location include/gpsd.h))/.. && $$ROOT/$(execpath :maskaudit) -c . > $$OUT",
+)
+
+genrule(
+    name = "ais_json_src",
+    tools = [":jsongen"],
+    outs = ["ais_json.i"],
+    cmd = "$(execpath :jsongen) --ais --target=parser > $@",
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
+    name = "gps",
+    srcs = [
+        "libgps/ais_json.c",
+        "libgps/bits.c",
+        "libgps/gpsdclient.c",
+        "libgps/gpsutils.c",
+        "libgps/hex.c",
+        "libgps/json.c",
+        "libgps/libgps_core.c",
+        "libgps/libgps_dbus.c",
+        "libgps/libgps_json.c",
+        "libgps/libgps_shm.c",
+        "libgps/libgps_sock.c",
+        "libgps/netlib.c",
+        "libgps/ntpshmread.c",
+        "libgps/os_compat.c",
+        "libgps/rtcm2_json.c",
+        "libgps/rtcm3_json.c",
+        "libgps/shared_json.c",
+        "libgps/timespec_str.c",
+        ":gps_maskdump_src",
+    ],
+    hdrs = glob(["include/*.h"], exclude = ["include/libgpsmm.h"]),
+    textual_hdrs = [":ais_json_src"],
+    copts = ["-std=c99"],
+    defines = [
+        "AIVDM_ENABLE=1",
+        "GPSD_PROTO_VERSION_MAJOR=16",
+        "GPSD_PROTO_VERSION_MINOR=1",
+        "HAVE_ARPA_INET_H=1",
+        "HAVE_CFMAKERAW=1",
+        "HAVE_CLOCK_GETTIME=1",
+        "HAVE_DAEMON=1",
+        "HAVE_FCNTL=1",
+        "HAVE_NETDB_H=1",
+        "HAVE_NETINET_IN_H=1",
+        "HAVE_STDATOMIC_H=1",
+        "HAVE_STRNLEN=1",
+        "HAVE_STPNCPY=1",
+        "HAVE_SYS_SOCKET_H=1",
+        "HAVE_SYS_UN_H=1",
+        "HAVE_SYSLOG_H=1",
+        "MAX_CLIENTS=64",
+        "MAX_DEVICES=6",
+        "SOCKET_EXPORT_ENABLE=1",
+        "VERSION=\\\"3.27.5\\\"",
+    ],
+    includes = ["include"],
+    linkopts = [
+        "-lm",
+        "-pthread",
+    ],
+)
+
+cc_library(
+    name = "gpsmm",
+    hdrs = ["include/libgpsmm.h"],
+    srcs = ["libgps/libgpsmm.cpp"],
+    includes = ["include"],
+    deps = [":gps"],
+)
+
+alias(
+    name = "libgps",
+    actual = ":gps",
+    visibility = ["//visibility:public"],
+)
+
+alias(
+    name = "libgpsmm",
+    actual = ":gpsmm",
+    visibility = ["//visibility:public"],
+)
+
+cc_test(
+    name = "libgps_test",
+    srcs = ["tests/libgps_test.c"],
+    deps = [":libgps"],
+)
+
+cc_test(
+    name = "libgpsmm_test",
+    srcs = ["tests/libgpsmm_test.cc"],
+    deps = [":libgpsmm"],
+)

--- a/modules/libgps/3.27.5.bcr.1/overlay/include/gpsd_config.h
+++ b/modules/libgps/3.27.5.bcr.1/overlay/include/gpsd_config.h
@@ -1,0 +1,49 @@
+#ifndef GPSD_CONFIG_H
+#define GPSD_CONFIG_H
+
+#define VERSION "3.27.5"
+#define REVISION "release-3.27.5"
+#define GPSD_PROTO_VERSION_MAJOR 16
+#define GPSD_PROTO_VERSION_MINOR 1
+#define GPSD_URL "https://gpsd.io/"
+#define MAX_CLIENTS 64
+#define MAX_DEVICES 6
+
+#if !defined(_POSIX_C_SOURCE)
+#define _POSIX_C_SOURCE 200809L
+#endif
+
+#if !defined(_DEFAULT_SOURCE)
+#define _DEFAULT_SOURCE
+#endif
+
+#if !defined(_GNU_SOURCE)
+#define _GNU_SOURCE 1
+#endif
+
+#if !defined(_DARWIN_C_SOURCE)
+#define _DARWIN_C_SOURCE 1
+#endif
+
+#define AIVDM_ENABLE 1
+#define SOCKET_EXPORT_ENABLE 1
+
+#define HAVE_ARPA_INET_H 1
+#define HAVE_CFMAKERAW 1
+#define HAVE_CLOCK_GETTIME 1
+#define HAVE_DAEMON 1
+#define HAVE_FCNTL 1
+#define HAVE_NETDB_H 1
+#define HAVE_NETINET_IN_H 1
+#define HAVE_STDATOMIC_H 1
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
+#define HAVE_STRLCAT 1
+#define HAVE_STRLCPY 1
+#endif
+#define HAVE_STRNLEN 1
+#define HAVE_STPNCPY 1
+#define HAVE_SYS_SOCKET_H 1
+#define HAVE_SYS_UN_H 1
+#define HAVE_SYSLOG_H 1
+
+#endif

--- a/modules/libgps/3.27.5.bcr.1/overlay/tests/libgps_test.c
+++ b/modules/libgps/3.27.5.bcr.1/overlay/tests/libgps_test.c
@@ -1,0 +1,6 @@
+#include "gps.h"
+
+int main(void) {
+  const char* mask = gps_maskdump(0);
+  return (mask != 0) ? 0 : 1;
+}

--- a/modules/libgps/3.27.5.bcr.1/overlay/tests/libgpsmm_test.cc
+++ b/modules/libgps/3.27.5.bcr.1/overlay/tests/libgpsmm_test.cc
@@ -1,0 +1,16 @@
+#include <libgpsmm.h>
+
+int main() {
+  // GPSD_LOCAL_FILE makes libgps read a plain file instead of a daemon socket,
+  // so a real gpsmm object exercises the documented C++ API with no gpsd,
+  // socket or network involved.
+  gpsmm gps(GPSD_LOCAL_FILE, "/dev/null");
+  if (!gps.is_open()) {
+    return 1;  // constructor failed to open
+  }
+  if (nullptr == gps.stream(WATCH_ENABLE)) {
+    return 1;  // local-file streams are read-only, so this always succeeds
+  }
+  (void)gps.read();  // /dev/null is an immediate EOF
+  return 0;
+}

--- a/modules/libgps/3.27.5.bcr.1/presubmit.yml
+++ b/modules/libgps/3.27.5.bcr.1/presubmit.yml
@@ -1,0 +1,21 @@
+matrix:
+  platform:
+    - ubuntu2204
+    - macos
+    - macos_arm64
+  bazel:
+    - 8.x
+    - 9.x
+tasks:
+  verify_targets:
+    name: Verify build and test targets
+    platform: ${{ platform }}
+    bazel: ${{ bazel }}
+    build_targets:
+      - "@libgps//:libgps"
+      - "@libgps//:libgpsmm"
+      - "@libgps//:libgps_test"
+      - "@libgps//:libgpsmm_test"
+    test_targets:
+      - "@libgps//:libgps_test"
+      - "@libgps//:libgpsmm_test"

--- a/modules/libgps/3.27.5.bcr.1/source.json
+++ b/modules/libgps/3.27.5.bcr.1/source.json
@@ -1,0 +1,12 @@
+{
+    "type": "archive",
+    "url": "https://download-mirror.savannah.gnu.org/releases/gpsd/gpsd-3.27.5.tar.xz",
+    "integrity": "sha256-3Epiutg1KCuueIdyvHzI+L7Ex6SOjc7rN0d6iQkcRlY=",
+    "strip_prefix": "gpsd-3.27.5",
+    "overlay": {
+        "BUILD.bazel": "sha256-M2WjI7o5mvN6HdvYH/LMUgdiS83YWlEWVgtt6FsenIY=",
+        "include/gpsd_config.h": "sha256-4Ly2tC8b677XvKEdIhzXXjlWiRFzWk5CQQWCfFUyMBg=",
+        "tests/libgps_test.c": "sha256-wkzB9v6K1AzmkrcEABJUTjAZzQQSGtAMFVdUkqT3bt4=",
+        "tests/libgpsmm_test.cc": "sha256-TphhflyDMbrHZMNmdoZkICgbRkx8jWU0hHifspjRmwM="
+    }
+}

--- a/modules/libgps/metadata.json
+++ b/modules/libgps/metadata.json
@@ -11,7 +11,8 @@
         "https://download-mirror.savannah.gnu.org/releases/gpsd/"
     ],
     "versions": [
-        "3.27.5"
+        "3.27.5",
+        "3.27.5.bcr.1"
     ],
     "yanked_versions": {}
 }

